### PR TITLE
[job submission] Prefix job ID with `raysubmit_` and pass `job_name` metadata

### PR DESCRIPTION
**Pull request by edoakes**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/job/job_manager.py [hunk 1]
```diff
@@ -25,6 +27,20 @@
     create_task = asyncio.ensure_future
 
 
+def generate_job_id() -> str:
```
Comment (tchordia) [inline dashboard/modules/job/job_manager.py:30]: we probably want to make this a lil bit more generic if we ever want to reuse the part where we are generating characters

File: dashboard/modules/job/job_manager.py [hunk 2]
```diff
@@ -67,7 +83,10 @@ def __init__(self, job_id: str, metadata: Dict[str, str]):
         self._runtime_env = ray.get_runtime_context().runtime_env
 
         self._metadata = metadata
-        self._metadata[JOB_ID_METADATA_KEY] = job_id
+        if JOB_ID_METADATA_KEY not in self._metadata:
```
Comment (tchordia) [inline dashboard/modules/job/job_manager.py:86]: I think a better pattern here may be to use dict update notation + a default metadata dict. so it's obvious what's going on
```
self._metadata = {
** default_metadata,
** passed_metadata
}
```

File: dashboard/modules/job/job_manager.py [hunk 3]
```diff
@@ -67,7 +83,10 @@ def __init__(self, job_id: str, metadata: Dict[str, str]):
...
+            self._metadata[JOB_ID_METADATA_KEY] = job_id
+        if JOB_NAME_METADATA_KEY not in self._metadata:
+            self._metadata[JOB_NAME_METADATA_KEY] = job_id
```
Comment (jiaodong) [inline dashboard/modules/job/job_manager.py:89]: how we're planning to use this field if we assign job_id to it, same as `JOB_ID_METADATA_KEY` ?

File: dashboard/modules/job/tests/test_job_manager.py [hunk 4]
```diff
@@ -238,11 +253,24 @@ def dict_to_str(d):
         wait_for_condition(
             check_job_succeeded, job_manager=job_manager, job_id=job_id)
         assert dict_to_str({
+            JOB_NAME_METADATA_KEY: job_id,
             JOB_ID_METADATA_KEY: job_id,
             "key1": "val1",
             "key2": "val2"
         }) in job_manager.get_job_logs(job_id)
```
Comment (jiaodong) [inline dashboard/modules/job/tests/test_job_manager.py:260]: I think it's worth considering adding a "get_metadata" API for jobs in this case, if user wants to submit a job with tags for experiment management / filtering.

**Comment by jiaodong** [comment]
Implementation looks good to me, just some comments and there's a failing test on CI. Feel free to ping me if you need to merge this asap to unblock.

File: dashboard/modules/job/job_manager.py (hunk 1 above)
Comment (edoakes) [reply dashboard/modules/job/job_manager.py:30]: unlikely, can do that if it's needed

File: dashboard/modules/job/job_manager.py (hunk 2 above)
Comment (edoakes) [reply dashboard/modules/job/job_manager.py:86]: done

**Review by jiaodong — APPROVED** [review]

**Pull request merged**